File: java/src/main/java/org/example/crypto/Ed25519Example.java
package org.example.crypto;

import java.nio.charset.StandardCharsets;
import java.security.*;

public final class Ed25519Example {
    private Ed25519Example() {}

    public static KeyPair generateKeyPair() throws Exception {
        return KeyPairGenerator.getInstance("Ed25519").generateKeyPair();
    }

    public static byte[] sign(PrivateKey key, byte[] message) throws Exception {
        Signature signature = Signature.getInstance("Ed25519");
        signature.initSign(key);
        signature.update(message);
        return signature.sign();
    }

    public static boolean verify(PublicKey key, byte[] message, byte[] signatureValue) throws Exception {
        Signature signature = Signature.getInstance("Ed25519");
        signature.initVerify(key);
        signature.update(message);
        return signature.verify(signatureValue);
    }

    public static void main(String[] args) throws Exception {
        KeyPair keys = generateKeyPair();
        byte[] message = "signed message".getBytes(StandardCharsets.UTF_8);
        byte[] signature = sign(keys.getPrivate(), message);
        CryptoSupport.require(verify(keys.getPublic(), message, signature), "Ed25519 failed");
        System.out.println("Ed25519Example signature=" + CryptoSupport.hex(signature));
    }
}

File: python/ed25519_example.py
"""Ed25519 signing and verification."""

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric import ed25519


def sign(private_key: ed25519.Ed25519PrivateKey, message: bytes) -> bytes:
    return private_key.sign(message)


def verify(public_key: ed25519.Ed25519PublicKey, message: bytes, signature: bytes) -> bool:
    try:
        public_key.verify(signature, message)
        return True
    except InvalidSignature:
        return False


def main() -> None:
    private_key = ed25519.Ed25519PrivateKey.generate()
    message = b"signed message"
    signature = sign(private_key, message)
    assert verify(private_key.public_key(), message, signature)
    assert not verify(private_key.public_key(), message + b"!", signature)
    print(f"Ed25519Example signature={signature.hex()}")


if __name__ == "__main__":
    main()
